src/app.ts

```diff
@@ -86,6 +109,132 @@ const cachingMiddleware = apicache.options({
   }
 }).middleware;
 
+const SAML_NAMESPACE = {
+  ASSERTION: "urn:oasis:names:tc:SAML:2.0:assertion",
+  PROTOCOL: "urn:oasis:names:tc:SAML:2.0:protocol"
+};
+
+const getFiscalNumberFromPayload = (doc: Document): Option<string> => {
+  return fromNullable(
+    doc.getElementsByTagNameNS(SAML_NAMESPACE.ASSERTION, "Attribute")
+  )
+    .mapNullable(collection => {
+      return tryCatch(() => {
+        return Array.from(collection).filter(
+          elem => elem.getAttribute("Name") === "fiscalNumber"
+        )[0];
+      });
+    })
+    .mapNullable(maybeElem => maybeElem.toNullable())
+    .mapNullable(_ => _.textContent?.trim());
+};
+
+const getRequestIDFromPayload = (tagName: string, attrName: string) => (
+  doc: Document
+): Option<string> => {
+  return fromNullable(
+    doc.getElementsByTagNameNS(SAML_NAMESPACE.PROTOCOL, tagName).item(0)
+  ).chain(element =>
+    fromEither(NonEmptyString.decode(element.getAttribute(attrName)))
+  );
+};
+
+const getRequestIDFromRequest = getRequestIDFromPayload("AuthnRequest", "ID");
+const getRequestIDFromResponse = getRequestIDFromPayload(
+  "Response",
+  "InResponseTo"
+);
+
+const SpidBaseMsg = t.interface({
+  // Timestamp of Request/Response creation
+  createdAt: UTCISODateFromString,
+
+  // Date of the SPID request / response in YYYY-MM-DD format
+  createdAtDay: PatternString("^[0-9]{4}-[0-9]{2}-[0-9]{2}$"),
+
+  // IP of the client that made a SPID login action
+  ip: IPString,
+
+  // XML payload of the SPID Request/Response
+  payload: t.string,
+
+  // Payload type: REQUEST or RESPONSE
+  payloadType: t.keyof({ REQUEST: null, RESPONSE: null })
+});
+
+export type SpidBaseMsg = t.TypeOf<typeof SpidBaseMsg>;
+
+const callback = (
+  sourceIp: string | null,
+  payload: string,
+  payloadType: "REQUEST" | "RESPONSE"
+): void => {
+  const maybeXmlPayload = tryCatch(() =>
+    new DOMParser().parseFromString(payload, "text/xml")
+  );
+  if (isNone(maybeXmlPayload)) {
+    log.error(`SpidLogCallback|ERROR=Cannot parse SPID XML Payload`);
+    return;
+  }
+
+  const doc = maybeXmlPayload.value;
+  const requestId =
+    payloadType === "REQUEST"
+      ? getRequestIDFromRequest(doc)
+      : getRequestIDFromResponse(doc);
+  const tFiscalCode = getFiscalNumberFromPayload(doc);
+
+  if (requestId.isNone()) {
+    log.error(
+      `SpidLogCallback| ERROR | Cannot recognize Request ID on XML Payload`
+    );
+    return;
+  }
+
+  if (tFiscalCode.isNone() && payloadType === "RESPONSE") {
+    log.error(
+      `SpidLogCallback| ERROR | Cannot recognize fiscal Code on XML Payload provided by SAMLResponse`
+    );
+    return;
+  }
+  const item = {
+    fiscalCode: tFiscalCode.toUndefined(),
+    spidRequestId: requestId.toUndefined()
+  };
+
+  const maybeBaseMsg = SpidBaseMsg.decode({
```

Comment:
there's no need to call decode, just type the object

```suggestion
  const maybeBaseMsg : SpidBaseMsg = {
```